pkg/audit/reporter.go

```diff
@@ -0,0 +1,96 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package audit
+
+import (
+	"context"
+
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/provider"
+	dbx "storj.io/storj/pkg/statdb/dbx"
+	proto "storj.io/storj/pkg/statdb/proto"
+	"storj.io/storj/pkg/statdb/sdbclient"
+)
+
+type reporter interface {
+	RecordFailedAudits(ctx context.Context, failedNodes []*pb.Node) (err error)
+}
+
+// Reporter records audit reports in statdb and implements the reporter interface
+type Reporter struct {
+	statdb sdbclient.Client
+}
+
+var (
+	port   = ":7777"
+	apiKey = []byte("")
+)
+
+// NewReporter instantiates a reporter
+func NewReporter(ctx context.Context) (reporter *Reporter, err error) {
+	ca, err := provider.NewCA(ctx, 12, 14)
+	if err != nil {
+		return nil, err
+	}
+	identity, err := ca.NewIdentity()
+	if err != nil {
+		return nil, err
+	}
+	client, err := sdbclient.NewClient(identity, port, apiKey)
+	if err != nil {
+		return nil, err
+	}
+	return &Reporter{statdb: client}, nil
+}
+
+// RecordFailedAudits saves failed audit details to statdb
+func (reporter *Reporter) RecordFailedAudits(ctx context.Context, failedNodes []*pb.Node) (err error) {
+	nodes, err := reporter.recordEntryIfNotExists(ctx, failedNodes)
+	if err != nil {
+		return err
+	}
+
+	finished := false
+	retries := 0
+	for !finished && retries < 3 {
+		_, failedNodes, err := reporter.statdb.UpdateBatch(ctx, nodes)
+		if err != nil {
+			return err
+		}
+		if len(failedNodes) == 0 {
+			finished = true
+		}
+		nodes = failedNodes
+		retries++
+	}
+	if retries == 3 {
```

Comment:
`if retries == 3 && len(nodes) > 0`. It's possible to have exactly three retries and have all the nodes get properly updated

Reply:
Also you don't need to but I feel more comfortable with `retries >= 3` rather than `retries == 3`

Additionally, you should define a constant like `MAX_RETRIES = 3` somewhere and use that instead of magic numbering it (you are using `3` in multiple places so it makes sense to replace it with a constant)

Reply:
ended up adding this to the Reporter struct and Config